### PR TITLE
[M] CANDLEPIN-1195: Added catch for EntityNotFoundException

### PR DESCRIPTION
**Pull request by joshmalbrecht**
- Added a catch block for ConsumerCurator.getConsumer() so that concurrently deleted entities that are eagerly fetched for Consumer does not cause a 500 error.

This change fixes the following spec test that intermittently fails:
`ConsumerResourceSpecTest.shouldReturn404Or410WhenConcurrentRegisterIsDeletedByAnotherRequest`

### Testing

You can add the `@RepeatedTest(100)` to the problematic `ConsumerResourceSpecTest.shouldReturn404Or410WhenConcurrentRegisterIsDeletedByAnotherRequest` spec test and compare before and after the changes.

### CODE REVIEW

File: src/test/java/org/candlepin/model/ConsumerCuratorTest.java [hunk 1]
```diff
@@ -332,6 +338,22 @@ public void testGetConsumerInputPartitioning() {
...
+        doReturn(spyEntityManager).when(spyCurator).getEntityManager();
+
+        QueryImplementor<Consumer> mockQuery = Mockito.mock(QueryImplementor.class);
+        doReturn(mockQuery).when(spyEntityManager).createQuery(any(CriteriaQuery.class));
+        doThrow(EntityNotFoundException.class).when(mockQuery).getSingleResult();
+
+        Consumer actual = spyCurator.getConsumer("uuid");
+
+        assertNull(actual);
+    }
```
[the remark this answers is not in the document]
Comment (joshmalbrecht) [reply src/test/java/org/candlepin/model/ConsumerCuratorTest.java:355]: Updated

File: src/test/java/org/candlepin/model/ConsumerCuratorTest.java [hunk 2]
```diff
@@ -332,6 +338,22 @@ public void testGetConsumerInputPartitioning() {
         }
     }
 
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testGetConsumerReturnsNullWhenEntityNotFoundExceptionIsThrown() {
```
Comment (bakajstep) [inline src/test/java/org/candlepin/model/ConsumerCuratorTest.java:343]: Why is all the mocking here? This is curator, so it has `DatabaseTestFixture` so it could be only this:
```
@Test
public void testGetConsumerReturnsNullWhenEntityNotFoundExceptionIsThrownV2() {
    Consumer consumer = consumerCurator.getConsumer("unknown-uuid");

    assertNull(consumer);
}
```

**Review by bakajstep — COMMENTED** [review]
i have one question otherwise LGTM

File: src/test/java/org/candlepin/model/ConsumerCuratorTest.java (hunk 2 above)
Comment (joshmalbrecht) [reply src/test/java/org/candlepin/model/ConsumerCuratorTest.java:343]: The problem is that `TypedQuery.getSingleResult()` would throw a `NoResultException` instead of the `EntityNotFoundException`.

Comment (Ceiu) [reply src/test/java/org/candlepin/model/ConsumerCuratorTest.java:343]: We need a test that checks not the not-found path, but "is kind of found, but base entity no longer exists" -- which is what triggers EntityNotFoundException rather than NoResultException.

Mocking isn't ideal here, but it makes sense given the specific type of exception we're targeting. We'd need to prime Hibernate with partial state -- a consumer object that internally thinks it has some children, but that don't actually exist. I don't know how possible that'd be without mocks or reflection.

File: src/main/java/org/candlepin/model/ConsumerCurator.java [hunk 3]
```diff
@@ -543,6 +544,10 @@ public Consumer getConsumer(String uuid) {
         catch (NoResultException e) {
             return null;
         }
+        catch (EntityNotFoundException e) {
```
Comment (Ceiu) [inline src/main/java/org/candlepin/model/ConsumerCurator.java:547]: Rare agreement with sonar. We can combine these two branches since it's effectively the same thing for us here.

```
catch (EntityNotFoundException | NoResultException e) {
   ...
}
```

**Review by Ceiu — CHANGES_REQUESTED** [review]
The major question I have here is whether or not this is the right answer. EntityNotFound is kind of a weird exception in that it means we have a murky state within Hibernate with partial data. Are we sure that we want to hide that erroneous state?

File: src/test/java/org/candlepin/model/ConsumerCuratorTest.java (hunk 2 above)
Comment (bakajstep) [reply src/test/java/org/candlepin/model/ConsumerCuratorTest.java:343]: ohh ok I misinterpreted the EntityNotFoundException

File: src/main/java/org/candlepin/model/ConsumerCurator.java (hunk 3 above)
Comment (Ceiu) [reply src/main/java/org/candlepin/model/ConsumerCurator.java:547]: Alternatively, as Sourcery suggests, keep both branches, but add warn/debug/error logging in the EntityNotFound branch since it's exceptional and weird.

**Comment by joshmalbrecht** [comment]
> The major question I have here is whether or not this is the right answer. EntityNotFound is kind of a weird exception in that it means we have a murky state within Hibernate with partial data. Are we sure that we want to hide that erroneous state?

@Ceiu  Yeah, that's a good point, and I'm not so sure that returning null from the curator and hiding/logging the error is the best answer as well. Since the exception is more broad than this specific edge case, we would be covering up a much larger problem.

What are your thoughts on tossing out this specific edge case from the problematic test by adding a condition that checks for a 500 status code and checks for a specific line in the error message? 

Something like the following:

```java
                try {
                    userClient.consumers().deleteConsumer(consumer.getUuid());
                }
                catch (ApiException e) {
                    if (e.getCode() == 410 || e.getCode() == 404) {
                        expectedExceptions.add(e);
                    }
                    else if (e.getCode() == 500 &&
                        e.getMessage().contains("Unable to find org.candlepin.model.IdentityCertificate")) {

                        // While concurrently deleting and retrieving a Consumer, it is possible that we run
                        // into a EntityNotFoundException because the IdentityCertificate is eagerly loaded.
                        // Since this error does not pertain to the objective of this test, we are throwing
                        // out this exception.
                    }
                    else {
                        unexpectedExceptions.add(e);
                    }
                }
```

**Comment by Ceiu** [comment]
Hrmm... I'm not sure I like that, either, since we're intentionally surfacing 500s to the user. Maybe keeping the structure you have now with ENF having its own catch branch and then doing a `log.error` so we can see how frequently it's happening in the logs. It's mostly hidden in terms of how it functions, but we'll still find out if it's a common enough problem to hit this.

Given how the test is operating, this specific instance for it seems to be expected or even intended to some degree. That feels a bit off, but I'm not sure the archeology necessary to find a better answer would give us anything new.

File: src/main/java/org/candlepin/model/ConsumerCurator.java (hunk 3 above)
Comment (joshmalbrecht) [reply src/main/java/org/candlepin/model/ConsumerCurator.java:547]: Kept both branches, and added an error log to the `EntityNotFound` branch.

**Review by Ceiu — APPROVED** [review]
